### DISCUSSION TITLE
Add more than 8 relay support in knx menu

### DISCUSSION THREAD

**Discussion by noroozihossein** [ascillato/Tasmota_KNX#583] [Ideas]
@ascillato hello dear Adrian hope that you’re well
First of all many thanks to developing tasmota KNX it’s very useful in projects, 
I have a question, i use tasmota32 on esp32 , as you know we can have more than 8 gpio to relate to relays and ... , is there anyway to add more than 8 relays in tasmota KNX web interface? Otherwise we should use another esp running tasmota to have another 8 relays to setup, 
I will be appreciated if you can help thank you

**Comment by ascillato** [comment]
Hi,

Yes, you are right. I will add support for more relays for ESP32.

I'm refactoring the whole driver so as to support more features (that will take some time), but I think I can do a fast patch for the actual version for ESP32.

Give some days that I'm in the middle of a project. After that I should have some free time to add that. Thanks.

**Comment by noroozihossein** [comment]
Hi dear friend 
Very good so I’m waiting to your confirm, also we have two models of ESP32 , normal model has 30 pin and we have another model with 38pin so we could have more GPIOs, i use ESP32 38 pin to control KNX actuators or using ESP32 38 pin with relays and control via KNX touch interface, so if you can add ability to control at least 16 relays instead of 8 relays, it was so useful ! I can help you in KNX side if you want any information, 
Thank you friend.

**Comment by noroozihossein** [comment]
@ascillato hi dear friend 
Some suggestions and ideas to more powerful than KNX version :
1) Add button with selectable functions: toggle, 1 , 0 or special value to special group address ( using tasmota device as binary input to command and communicate with other KNX TP devices)
2) Add single and double and hold function for each button and selectable : toggle, 1 , 0 or special value to send to a group address ( usable for calling a scene in KNX protocol or changing single actuator)
3) Add at least 16 relay and 16 button to use in KNX web UI
4) As you maybe know in KNX protocol voice Assistant is a unique option and has more cost to implement, we use and simulate binary to control KNX actuators and also use them via HUE Bridge to use Via Alexa, i saw at the present just able to have 8 device in Hue emulator if it possible add more devices,(ESP32 Edition)
5) maybe it will possibly to add thermostat functionality in KNX in future using ESP32 ( just use 3 or 4 relays with change over function (interlock) to control low, medium,high and valve, controlling via one TP KNX interface same thermostatic keys.
5) relative and absolute dimming and also shutter and blinds control that i saw you’re planning to add number 5 functions to next refactoring version of tasmota KNX
Thank you

**Reply by ascillato** [reply]
Hi,

So far I was very busy but I didn't forget about this. Below I'm answering your proposals:

>1. Add button with selectable functions: toggle, 1 , 0 or special value to special group address ( using tasmota device as binary input to command and communicate with other KNX TP devices)
>2. Add single and double and hold function for each button and selectable : toggle, 1 , 0 or special value to send to a group address ( usable for calling a scene in KNX protocol or changing single actuator)

That is already possible by using rules. You can set the behavior you want for a button with Tasmota commands and then send the value you need using KNXTX_** commands.

>3. Add at least 16 relay and 16 button to use in KNX web UI

Yes, will do that as a patch as soon as I'm able to.

>4. As you maybe know in KNX protocol voice Assistant is a unique option and has more cost to implement, we use and simulate binary to control KNX actuators and also use them via HUE Bridge to use Via Alexa, i saw at the present just able to have 8 device in Hue emulator if it possible add more devices,(ESP32 Edition)

This should be posted as a feature request directly to Tasmota discussions as it is about the HUE Bridge emulation.

>5. maybe it will possibly to add thermostat functionality in KNX in future using ESP32 ( just use 3 or 4 relays with change over function (interlock) to control low, medium, high and valve, controlling via one TP KNX interface same thermostatic keys.

That is already possible with Tasmota. You can use rules or the Thermostat feature of Tasmota.

>6. relative and absolute dimming and also shutter and blinds control that i saw you’re planning to add number 5 functions to next refactoring version of Tasmota KNX

Yes, color and dimming support is part of the idea for the refactoring. So far, shutters and blinds are not part of the plan for now, Sorry.
I don't have a release time for all this as I'm doing it in my free time that at this moment is scarce, sorry. But as soon as this is ready I will tell you here.

**Reply by noroozihossein** [reply]
@ascillato Hello dear friend Adrian hope that you’re well, since Tasmota 10.0 released, I thought that also you refactored KNX section to support more than 8 relay and button support in GUI , but it seems nothing changes, still don’t have enough time to work on it?

**Comment by noroozihossein** [comment]
@ascillato Hello dear friend Adrian hope that you’re well, since Tasmota 10.0 released, I thought that also you refactored KNX section to support more than 8 relay and button support in GUI , but it seems nothing changes, still don’t have enough time to work on it?

**Reply by ascillato** [reply]
Hi, I have been very busy at work. Sorry, I didn't have any free time to work on this.

**Reply by noroozihossein** [reply]
Nice job friend ok so please whenever you had free time and could work on it inform us to test
Thank you.

**Comment by AhmedOsama** [comment]
@ascillato thank you very much for everything, please do let us know when you add the patch for 16 relay and 16 button to use in KNX web UI 

I know you do this in your free time it is much appreciated thank you very much

**Comment by AhmedOsama** [comment]
@noroozihossein were you able to find a work around for adding more than 8 relays ? Using knx ?

**Reply by noroozihossein** [reply]
yes i use and integrate KNX devices normally in my projects now i find new issue in Tasmota32 KNX section
in Tasmota ESP8266 KNX version everything works normally one time telegram sending from BUS is ok and in Tasmota side module get and act without problem but in Tasmota32 firmware and modules need to send several times Telegrams from BUS to get and act in Tasmota32 module side , i tested with several modules with different Tasmota32 version Firmware also it seems this issue related to Tasmota32 KNX driver. @ascillato

**Comment by AhmedOsama** [comment]
@noroozihossein if that is the case then I believe you should open an issue so that devs can pick it up by my question is are you able to use 16 relays for esp32 with knx

**Reply by noroozihossein** [reply]
As i use Tasmota KNX GUI to integrate with KNX device so the answer is unfortunately not , maybe it is possible to relate more than 8 relays with KNX group address via rules but I don’t know how to , it were supposed to friend @ascillato refactor the KNX driver to support it in KNX GUI section but he still had not enough extera time to work on it

**Comment by noroozihossein** [comment]
@ascillato hello dear friend hope that you’re well,it seems you still doesn’t have enough time to refactor firmware to support 16 relays am I right?and also dimmer support for knx

**Reply by ascillato** [reply]
Hi, sorry for the long wait. I will try to allocate some time in any of the following weekends for adding 16 relays support.

**Reply by AhmedOsama** [reply]
Thank you, please do update when done

**Comment by noroozihossein** [comment]
I will be appreciated dear friend 🙏 please update me whenever you done it 
Thank you

**Comment by noroozihossein** [comment]
@ascillato hello dear friend how are you , hope that you are well , any news to update? still you did not have any free time to work on KNX section? let us know new updates in KNX section please 
thank you in advance.

**Comment by noroozihossein** [comment]
@ascillato how are you dear friend , is there any update?
